### PR TITLE
Added wait_for_lock optional argument

### PR DESCRIPTION
**Pull request by tdunning**
This extension should cause no impact on existing code.

### CODE REVIEW

**Comment by tdunning** [comment]
See issue #8

**Comment by tdunning** [comment]
Testing on 1.0 seems to fail due to timing issues caused by slow compilation. Running the tests a second time results in complete success. I have opened up some of the tolerances to deal with this.

**Review by vtjnash — COMMENTED** [review]
SGTM

Can you relax the timeouts in the tests? They are unacceptably tight right now, and should not be using fractional tolerances of less than 1 second.

What do you think of calling this kwarg `wait::Bool=true` (e.g. similar to `run`)?

**Comment by tdunning** [comment]
I would be happy to relax the timeouts in the test.

And any suggestions on API would be a great thing. I am relatively new to
Julia and often write non-idiomatic code as a result.

Where specifically are you looking for that kwarg?

**Comment by vtjnash** [comment]
The API looked good overall. I am proposing naming `wait_for_lock` as just `wait`. I am not sure the PidLockFailedError is needed either, and think that we can just use `error("Failed to get lock file $(repr(path)).")` here.

**Comment by tdunning** [comment]
Ah... that makes sense.

I will revise and push.

**Comment by tdunning** [comment]
Revised the pull request.

File: src/Pidfile.jl [hunk 1]
```diff
@@ -170,20 +175,40 @@ end
...
+                try
+                    rm(path)
+                catch ex
+                    isa(ex, IOError) || rethrow(ex)
+                end
+            end
+            file = tryopen_exclusive(path, mode)
+        end
+        if file === nothing
+            throw(PidLockFailedError("Failed to get lock file"))
```
Comment (vtjnash) [inline src/Pidfile.jl:205]: suggestion:
```suggestion
            error("Failed to get pidfile lock for $(repr(path)).")
```

**Review by vtjnash — COMMENTED** [review]
Also need to change the calls to `wait` to call instead `Pidfile.wait` in this function now

**Comment by tdunning** [comment]
Sorry ... I don't understand this comment about calling wait.

I don't see a definition of Pidfile.wait. Am I missing something? (well, I
am obviously missing something)

File: src/Pidfile.jl [hunk 2]
```diff
@@ -29,9 +30,13 @@ Optional keyword arguments:
      The file won't be deleted until 25x longer than this if the pid in the file appears that it may be valid.
      By default this is disabled (`stale_age` = 0), but a typical recommended value would be about 3-5x an
      estimated normal completion time.
+ - `wait`: If true, block until we get the lock, if false, throw PidLockFailedError if lock fails.
 """
 function mkpidlock end
 
+struct PidLockFailedError <: Exception
+    msg::String
+end
```
Comment (vtjnash) [inline src/Pidfile.jl:39]: also will need to update the test

```suggestion
 - `wait`: If true, block until we get the lock, if false, throw an exception if lock fails.
"""
function mkpidlock end

```

**Comment by vtjnash** [comment]
You cannot have a both a value and a function of the same name, so you need to qualify the function with the name of the global module

**Comment by tdunning** [comment]
I am on the test failures. Almost certainly my fault.

File: src/Pidfile.jl [hunk 3]
```diff
@@ -20,7 +20,8 @@ using Base.Sys: iswindows
 
 Create a pidfile lock for the path "at" for the current process
 or the process identified by pid or proc. Can take a function to execute once locked,
-for usage in `do` blocks, after which the lock will be automatically closed.
+for usage in `do` blocks, after which the lock will be automatically closed. If the lock fails
+and `wait` is false, then PidLockFailedError is thrown.
```
Comment (vtjnash) [inline src/Pidfile.jl:24]:
```suggestion
and `wait` is false, then an error is thrown.
```

**Comment by tdunning** [comment]
So we are back to the question of code coverage. I will take a look to see if there is a quick win on that.

**Comment by tdunning** [comment]
Yeah... so I am finding it difficult to create a repeatable test that munches on the mtime of the sentinel file.

The regression in coverage is one line that puts out a warning if the clock is skewed oddly (i.e. it finds a file from the future).

We have a few options here:

a) just drop the warning (I don't really know that it adds much value)

b) just take the hit in coverage

c) keep slogging on the test trying to figure out what is happening to make it non-deterministic.

These are in my preference since I am out of time for this.

Thoughts?

**Comment by vtjnash** [comment]
That doesn't sound related to your patch, so don't worry about it. Can you finish fixing the docs for the error, and I'll merge?

**Comment by tdunning** [comment]
Will do

File: src/Pidfile.jl [hunk 4]
```diff
@@ -147,9 +143,8 @@ Helper function for `open_exclusive` for deciding if a pidfile is stale.
 """
 function stale_pidfile(path::String, stale_age::Real)
     pid, hostname, age = parse_pidfile(path)
-    if age < -stale_age
-        @warn "filesystem time skew detected" path=path
-    elseif age > stale_age
+    age < 0 && @warn "filesystem time skew detected" path=path
```
Comment (vtjnash) [inline src/Pidfile.jl:146]:
```suggestion
    age < -stale_age && @warn "filesystem time skew detected" path=path
```

**Pull request merged**